File: src/kernel/scheduler.zig
const std = @import("std");
const ArrayList = std.ArrayList;
const Kernel = @import("kernel.zig");

const Self = @This();

// substitute for real dynamic data structure
processes: [16]Process,
kernel_process: Process,
next_process_id: u64,
dispatch_spinlock: Kernel.Spinlock,

shutdown: bool,
panic: bool,

pub const Thread = struct
{
    process: *Process,
    temporary_address_space: ?*volatile Kernel.Memory.Space,
    is_page_generator: bool,
    is_kernel_thread: bool,
};

pub fn init(self: *Self) void
{
    // @Spinlock
    self.kernel_process = Process
    {
        .id = self.next_process_id,
        .virtual_memory_space = &Kernel.kernel_memory_space,
        .type = .kernel,
        .name = "kernel",
        .permissions = Process.Permissions.all,
    };

    self.next_process_id += 1;
}

pub const Process = struct
{
    id: u64,
    address_space: *Kernel.Memory.Space,
    permissions: u64,
    type: Type,
    name: []const u8, // @TODO: should the process own the memory for the name?

    const Permissions = enum(u64)
    {
        none = 0,
        const all: u64 = std.math.maxInt(u64);
    };

    const Type = enum
    {
        normal,
        kernel,
    };
};

pub fn prepare_process(self: *Self, comptime process_type: Process.Type) ?*Process
{
    if (self.shutdown) return null;

    var process =
        if (process_type == .kernel) &self.kernel_process
        else Kernel.Arch.CPU_stop();

    process.address_space = 
        if (process_type == .kernel)
            &Kernel.memory_space
        else
            Kernel.Arch.CPU_stop();

    process.id = @atomicRmw(u64, &self.next_process_id, .Add, 1, .SeqCst);
    //process.address_space.reference_count = 1;
    process.type = process_type;
    process.permissions = Process.Permissions.all;

    if (process_type == .kernel)
    {
        process.name = "Kernel";
    }

    return process;
}

File: src/kernel/scheduler.rs
use kernel::*;

pub struct Process
{
    pub address_space: memory::AddressSpace,
    pub id: u64,
    pub handle_count: u64,
    pub permissions: ProcessPermissions,
    pub kind: ProcessKind,
}

pub struct Scheduler
{
    pub dispatch_spinlock: Spinlock,
    pub next_process_id: AtomicU64,

    // @TODO: volatile
    pub started: bool,
    pub panic: bool,
    pub shutdown: bool,
}

pub enum ProcessKind
{
    normal,
    kernel,
    desktop,
}

pub struct Writer
{
    lock: VolatilePointer<WriterLock>,
    kind: bool,
}

#[derive(Copy, Clone, PartialEq)]
pub enum ThreadState
{
    active,
    waiting_for_mutex,
    waiting_for_event,
    waiting_for_writer_lock,
    terminated,
}

#[derive(Copy, Clone, PartialEq)]
pub enum ThreadTerminatableState
{
    invalid_TS,
    terminatable,
    in_syscall,
    user_block_request,
}

#[derive(Copy, Clone)]
pub struct ThreadVolatile
{
}

#[derive(Copy, Clone)]
pub struct ThreadWriterLock
{
    pub lock: VolatilePointer<WriterLock>,
    pub kind: bool,
}

pub const max_wait_count: usize = 8;

#[derive(Copy, Clone)]
pub struct ThreadEvent
{
    pub threads: VolatilePointer<LinkedItem<Thread>>,
    pub events: [VolatilePointer<Event>; max_wait_count],
    pub event_count: Volatile<u64>,
}

#[derive(Copy, Clone)]
pub struct ThreadBlocking
{
    pub mutex: VolatilePointer<Mutex>,
    pub writer: ThreadWriterLock,
    pub event: ThreadEvent,
}

#[derive(Copy, Clone)]
pub struct Thread
{
    pub in_safe_copy: bool,

    pub process: *mut Process,

    pub last_interrupt_timestamp: u64,
    pub is_kernel_thread: bool,
    pub is_page_generator: bool,

    pub state: Volatile<ThreadState>,
    pub terminatable_state: Volatile<ThreadTerminatableState>,
    pub executing: Volatile<bool>,
    pub terminating: Volatile<bool>,
    pub blocking: ThreadBlocking,

    pub temporary_address_space: VolatilePointer<memory::AddressSpace>
}

bitflags!
{
    pub struct ProcessPermissions: u64
    {
        const networking = 1 << 0;
        const process_create = 1 << 1;
        const process_open = 1 << 2;
        const screen_modify = 1 << 3;
        const shutdown = 1 << 4;
        const take_system_snapshot = 1 << 5;
        const get_volume_information = 1 << 6;
        const window_manager = 1 << 7;
        const posix_subsystem = 1 << 8;
    }
}

impl Process
{
    pub fn register(&mut self, kind: ProcessKind)
    {
        self.id = unsafe { kernel.scheduler.next_process_id.fetch_add(1, Ordering::SeqCst) };
        self.address_space.reference_count = 1;
        // list, table
        self.handle_count = 1;
        self.permissions = ProcessPermissions::all();
        self.kind = kind;
    }

    //pub fn spawn<'b>(process_kind: ProcessKind) -> Option<&'b mut Process>
    //{
        //if unsafe { kernel.scheduler.shutdown } { return None }

        //let process = match process_kind
        //{
            //ProcessKind::normal | ProcessKind::desktop =>
            //{
                //unimplemented!();
            //},
            //ProcessKind::kernel =>
            //{
                //let result = &mut unsafe { kernel.process };
                //result
            //}
        //};
    //}
}
